File: src/dotnet/AzureDataFactory.TestingFramework/Models/Activities/ControlActivity/ForEachActivity.cs
// Copyright (c) Microsoft Corporation.
// Licensed under the MIT License.

using AzureDataFactory.TestingFramework.Models.Activities.Base;
using AzureDataFactory.TestingFramework.Models.Base;
using AzureDataFactory.TestingFramework.Models.Pipelines;

namespace AzureDataFactory.TestingFramework.Models;

public partial class ForEachActivity : IIterationActivity
{
    private List<string>? _items;
    public List<string> IterationItems => _items ?? throw new InvalidOperationException("Items have not been evaluated yet.");
    public override DataFactoryEntity Evaluate(PipelineRunState state)
    {
        _items = Items.Evaluate<List<string>>(state);

        return base.Evaluate(state);
    }

    internal override IEnumerable<PipelineActivity> EvaluateControlActivityIterations(PipelineRunState state, EvaluateActivitiesDelegate evaluateActivities)
    {
        // Note: using enumerator to support yield return in foreach
        using var enumerator = IterationItems.GetEnumerator();
        while (enumerator.MoveNext())
        {
            var scopedState = state.CreateIterationScope(enumerator.Current);
            foreach (var activity in evaluateActivities(Activities.ToList(), scopedState))
                yield return activity;

            state.AddScopedActivityResultsFromScopedState(scopedState);
        }
    }
}

File: src/python/azure_data_factory_testing_framework/models/activities/control_activities/for_each_activity.py
from typing import Callable, Generator

from azure_data_factory_testing_framework.generated.models import Activity, ControlActivity, ForEachActivity
from azure_data_factory_testing_framework.state import PipelineRunState


class ForEachActivity:
    def evaluate(self: ForEachActivity, state: PipelineRunState) -> ForEachActivity:
        self.items.evaluate(state)

        return super(ControlActivity, self).evaluate(state)

    def evaluate_control_activity_iterations(
        self: ForEachActivity,
        state: PipelineRunState,
        evaluate_activities: Callable[[PipelineRunState], Generator[Activity, None, None]],
    ) -> Generator[Activity, None, None]:
        for item in self.items.evaluated:
            scoped_state = state.create_iteration_scope(item)
            for activity in evaluate_activities(self.activities, scoped_state):
                yield activity

            state.add_scoped_activity_results_from_scoped_state(scoped_state)
